### ISSUE TITLE
Fly-Pie gets confused with multiple cursors on different screens

### ISSUE THREAD

**Issue by wsippel** [Schneegans/Fly-Pie#234]
## Describe the Bug
I run a triple monitor setup, one of them is a pen display. When I use the pen and open Fly-Pie, the menu opens in the top left corner of the screen the mouse cursor is on, not under the pen cursor. If I move the mouse cursor to the pen display screen first, Fly-Pie picks up the pen cursor, but the menu opens in the top left corner of the pen display screen, then it visibly slides to the cursor position. I believe the issue is that the pen is treated as a second, independent cursor, and Fly-Pie gets confused if there are multiple cursors on different screens.

## Expected Behavior
Fly-Pie should always pick up the currently used pointer device.

## System
EndeavourOS
Fly-Pie 16
Gnome Shell 42.2

**Comment by Schneegans** [comment]
Hi there & thanks for the report! This is a pretty complex topic and it could be that there are some issues left. AFAIK, multi-cursor support only works on Wayland. So I guess you are already using Wayland?

**Comment by wsippel** [comment]
Yes, I forgot to mention in the report that I'm on Wayland. It's certainly a niche usecase and probably a difficult to reproduce issue, so I'll gladly help if there's anything I can do - Fly-Pie is already awesome as is (thanks, by the way!), and absolutely perfect for a pen display setup.

**Comment by Schneegans** [comment]
Mhh... do you have some JavaScript knowledge? I think it would be most fruitful to debug this on your end. But this will involve some coding and testing :wink:

**Comment by wsippel** [comment]
Not JavaScript, but I've done some stuff in PHP, C, C++ and Rust. I see that the relevant code is probably in Menu.js, and that you're specifically accounting for tablets on Wayland. I also see that you initialize at 0,0 if you don't have "real" coordinates, which explains what I see. But I don't see anything related to identifying the screen, and I don't really understand the Clutter side of things.

**Comment by Schneegans** [comment]
Sorry for not getting back to you more early! I think one thing you could try is the following:
* Clone the Fly-Pie repository ([here](https://github.com/Schneegans/Fly-Pie/blob/main/docs/installation.md#c-cloning-the-latest-version-with-git) is a guide).
* Edit `src/extension/Menu.js` to add some debug prints. I think first we should try logging your device type. For this, add `utils.debug(device.get_device_type());` around [line 111](https://github.com/Schneegans/Fly-Pie/blob/main/src/extension/Menu.js#L111).
* Install the extension (`make install`).
* Restart GNOME Shell. As you are on Wayland, you have to log out and log in again.
* Open a console and let `journalctl -f -o cat` running while interacting with Fly-Pie. You log messages will show up there. The possible device types are shown here: https://gjs-docs.gnome.org/clutter10~10_api/clutter.inputdevicetype Which one is used by your input device?

**Comment by wsippel** [comment]
That theory is out of the window, the tablet does identify correctly (device type 4). 

I wonder if the keyboard event to invoke Fly-Pie might be the problem? The debug log always shows device 1, then 0 (for the mouse) or 4 (for the tablet). Could _deviceChangedID try to get a cursor position from the keyboard device, as that's technically the "last-device-changed" as far as I understand?

**Comment by wsippel** [comment]
I added a few more debug lines and I think I found something. This is the code:
`      // If we have to open the menu at a secondary pointer (e.g. from a stylus), we use
      // this enter event to position the menu.
      if (
        event.type() == Clutter.EventType.ENTER &&
        this._lastNonPointerDevice &&
        this._initialRepositioningRequired
      ) {
        utils.debug(event.get_device().get_device_name());
        utils.debug(this._lastNonPointerDevice.get_device_name());
        if (
          event.get_device().get_device_name() ===
          this._lastNonPointerDevice.get_device_name()
        ) {
          this._setPosition(this._pointerPos[0], this._pointerPos[1], false);
          this._initialRepositioningRequired = false;
          utils.debug(this._pointerPos[0]);
          utils.debug(this._pointerPos[1]);
        }
      }`
This is the debug output if the mouse cursor is on a different screen:

> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 1
> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 4
> [flypie@schneegans.github.com/src/extension/Menu.js:181:15] Virtual pointer device for seat
> [flypie@schneegans.github.com/src/extension/Menu.js:182:15] OpenTabletDriver Virtual Artist Tablet
> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 1

And this is the output if it's on the same screen:

> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 1
> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 4
> [flypie@schneegans.github.com/src/extension/Menu.js:181:15] Virtual pointer device for seat
> [flypie@schneegans.github.com/src/extension/Menu.js:182:15] OpenTabletDriver Virtual Artist Tablet
> [flypie@schneegans.github.com/src/extension/Menu.js:181:15] OpenTabletDriver Virtual Artist Tablet
> [flypie@schneegans.github.com/src/extension/Menu.js:182:15] OpenTabletDriver Virtual Artist Tablet
> [flypie@schneegans.github.com/src/extension/Menu.js:189:17] 5886.3701171875
> [flypie@schneegans.github.com/src/extension/Menu.js:190:17] 1245.310302734375
> [flypie@schneegans.github.com/src/extension/Menu.js:118:15] 1

I have no idea why it goes through the first if condition twice, but the obvious problem seems to be the second condition failing because the names don't match, but I have no idea where the 'Virtual pointer for seat' is coming from. It's apparently an internal Wayland thing? A quick web search led me to https://wayland.app/protocols/wlr-virtual-pointer-unstable-v1

**Comment by wsippel** [comment]
Sorry for the spam. The "virtual pointer device" is the mouse, no idea why it's called that. The core issue, as far as I can tell, is that event.get_coords() only sees the tablet when the mouse cursor is on the same screen or while the pen cursor is moving. So I tried to always keep the last known pen cursor position in a separate variable, and that kinda works, but now I run into a problem with the _clampToMonitor function. Because of course layoutManager.currentMonitor doesn't care about the pen pointer and returns the monitor the actual mouse pointer is on instead, and I don't see a function to easily get a monitor by cursor coordinates.

**Comment by Schneegans** [comment]
I think I'll just wait a couple of hours and you'll fix it yourself :stuck_out_tongue: 

There's [`global.display.get_monitor_index_for_rect(rect)`](https://gjs-docs.gnome.org/meta9~9_api/meta.display#method-get_monitor_index_for_rect). Maybe this could do the trick? The rect could be created with something along the lines of

```js
let rect = new Meta.Rectangle({
  x: ...,
  y: ...,
  width: 1,
  height: 1
});
```

**Comment by wsippel** [comment]
Good call. That seems to work, returned the correct index for a bit, but then it completely crashed my PC, requiring a hard reboot. Twice. Insane log spam, too. I guess it was trying to create tiny rectangles as fast as it could?

**Comment by wsippel** [comment]
No, I think my stupidity caused the crash. I tried to be clever and simply replaced `const monitor = Main.layoutManager.currentMonitor;` with `const monitor = this._monitor;`, where this._monitor is the index, but currentMonitor seems to be an object, and that was apparently enough to cause all sorts of weird behavior, up to and including complete system crashes. But now I'm kinda afraid to try random things. Do you think `const monitor = Main.layoutManager.monitors[this._monitor]` would work?

**Comment by wsippel** [comment]
You might have guessed it, but I tried anyway. It sort of works, and while it now clamps to the correct screen, it still displays on the wrong screen, and I'm completely out of ideas. Even hard coding the monitor only affects the clamping, not the display the menu actually shows up on. My best guess is that there's a get_pointer or currentMonitor or something somewhere else. Maybe not even in Menu.js.

**Comment by Schneegans** [comment]
Well, I guess the reason lies in the [`Background.open()`](https://github.com/Schneegans/Fly-Pie/blob/main/src/extension/Background.js#L122) method. It uses the current monitor (which apparently is determined via the mouse pointer only) to position the menu.

I think we will have to get rid of all calls to `currentMonitor` and try to manage it ourselves.

Good news is that I just dug out my old Wacom tablet and an external 7" screen. With both attached to my laptop, I can actually reproduce the issue! Bad news is that I'll be on vacation with the family for about four weeks from now, so my time to work on this will be somewhat limited.

If you want to mess around with the source code in the mean time - feel free to do so! A crashing session isn't too bad :). You can also run a nested GNOME Shell to quickly confirm that the code does not crash. This also speeds up development time as you do not have to log out / log in every time. Just run this:

```bash
cd <directory of Fly-Pie>
make install
MUTTER_DEBUG_DUMMY_MODE_SPECS=1600x900 dbus-run-session -- gnome-shell --nested --wayland
```
And a new window with a nested session will open. It's a bit difficult to Fly-Pie in there as the hot keys are captured by your main session. One approach is to configure a Touch Button for you menu and use this inside the nested session.

![image](https://user-images.githubusercontent.com/829942/179147027-a96e123a-3dbf-4510-a124-4a9b8131472b.png)

**Comment by Schneegans** [comment]
BTW, the initial menu movement from the top left corner to the pen position is a bug / regression. This must have been introduced with a recent change as I am pretty sure it wasn't there before... These are two different bugs I guess. But both need to be fixed!

**Comment by wsippel** [comment]
I looked into running Gnome nested, but it didn't seem super useful for debugging multi-monitor related problems. And good to hear that you could reproduce the issue, should make things a lot easier. I'll continue working on this the coming days. My code quality will probably suck, but maybe I can at least figure out what's going on and look for a feasible solution while you're gone. Enjoy your vacation! :smiley:

**Comment by Schneegans** [comment]
Yeah, I am using [clang-format](https://clang.llvm.org/docs/ClangFormat.html). You can either instruct your IDE to execute it whenever you save a file or you can run the script `scripts/clang-format.sh` occasionally. This will apply the style to all source files of the extension.

**Comment by Schneegans** [comment]
I finally revisited this issue and found a fix. The background now covers the entire stage (all monitors combined) - so there is no need for getting the correct monitor anymore. And I removed any initial movement animations. After some initial testing, it seems to work well.

If you want to, you could test the current `main` branch!

**Comment by wsippel** [comment]
Yes, works! And looks far more sane than what I tried to do. Thanks a lot!
